pgmpy/prediction/DynamicDMLRegressor.py

```diff
@@ -0,0 +1,851 @@
+                    f"Exactly one outcome variable must be defined in causal_graph. "
+                    f"Got {len(outcome_vars)}: {outcome_vars}"
+                )
+
+            # Store extracted roles
+            self.exposure_var_ = exposure_vars[0]
+            self.outcome_var_ = outcome_vars[0]
+            self.adjustment_vars_ = list(self.causal_graph.get_role("adjustment"))
+            self.pretreatment_vars_ = list(self.causal_graph.get_role("pretreatment"))
+            self.feature_columns_ = (
+                [self.exposure_var_] + self.adjustment_vars_ + self.pretreatment_vars_
+            )
+        else:
+            # No causal graph: use all features as confounders
+            self.exposure_var_ = None
+            self.outcome_var_ = None
+            self.adjustment_vars_ = []
+            self.pretreatment_vars_ = []
+            self.feature_columns_ = None
+
+        # STEP 3: Convert to numpy arrays
+        if hasattr(X, "columns"):
+            self.feature_names_in_ = np.array(X.columns)
+            X = X.values
+        else:
+            X = np.asarray(X)
+
+        T = np.asarray(T)
+        y = np.asarray(y)
+        groups = np.asarray(groups)
+
+        # Reshape T to 2D if 1D
+        if T.ndim == 1:
+            T = T.reshape(-1, 1)
+
+        self.groups_ = groups
+
+        # STEP 4: Infer n_periods if not provided
+        if self.n_periods is None:
+            self.n_periods_ = len(np.unique(groups))
+        else:
+            self.n_periods_ = self.n_periods
+
+        # STEP 5: Infer discrete_treatment and discrete_outcome if None
+        if self.discrete_treatment is None:
+            self.discrete_treatment_ = self._infer_discrete(T)
+        else:
+            self.discrete_treatment_ = self.discrete_treatment
+
+        if self.discrete_outcome is None:
+            self.discrete_outcome_ = self._infer_discrete(y)
+        else:
+            self.discrete_outcome_ = self.discrete_outcome
+
+        # STEP 6: Select models if 'auto'
+        self.model_y_ = self._select_model(self.model_y, self.discrete_outcome_)
+        self.model_t_ = self._select_model(self.model_t, self.discrete_treatment_)
+
+        # STEP 7: STAGE 1 - Cross-fitted nuisance estimation
+        self._fit_nuisances(X, T, y, groups)
+
+        # STEP 8: STAGE 2 - Backward recursive parameter estimation
+        self._fit_parameters()
+
+        # STEP 9: Compute asymptotic covariance
+        self._compute_covariance()
+
+        # STEP 10: Set sklearn-required attributes
+        self.n_features_in_ = X.shape[1]  # Number of features (columns)
+        self.n_samples_ = X.shape[0]  # Number of samples (rows)
+
+        # STEP 10: Return self
+        return self
+
+    def predict(self, X, T0=None, T1=None):
+        """
+        Predict treatment effects.
+
+        ALGORITHM:
+        Compute E[Y | do(T=T1)] - E[Y | do(T=T0)] using estimated coef_.
+        Effect = sum over periods t of coef_[t] @ (T1_t - T0_t)
```

Comment:
I don't think this is the correct algorithm for prediction. For prediction, we are trying to estimate: `E[Y | do(T=T_new), X_new]`. The equation here would instead compute the average treatment effect.